File: java/src/blog/recursive/DrawFractal.java
package blog.recursive;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.GeneralPath;
import java.util.concurrent.TimeUnit;

public class DrawFractal {
	private JFrame frame;
	private JPanel panel;

	public DrawFractal() {
		SwingUtilities.invokeLater(new Runnable() {
			public void run() {
				createAndShowGUI();
			}
		});
	}

	private void createAndShowGUI() {
		frame = new JFrame("Draw Fractal");
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		panel = new MyPanel();
		frame.add(panel);
		frame.pack();
		frame.setVisible(true);
	}

	class MyPanel extends JPanel {
		public MyPanel() {
			setBorder(BorderFactory.createLineBorder(Color.black));
		}

		@Override
		public Dimension getPreferredSize() {
			return new Dimension(500, 500);
		}
	}

	private void drawTriangle(double x, double y, double w, double h) {
		Graphics2D g = (Graphics2D) panel.getGraphics();

		double xp[] = {x, x+w, x+w/2, x};
		double yp[] = {y, y, y-h, y};

		GeneralPath polyline = new GeneralPath(GeneralPath.WIND_EVEN_ODD, 4);
		polyline.moveTo(xp[0], yp[0]);

		for (int i = 1; i < 4; i++)
			polyline.lineTo(xp[i], yp[i]);

		g.draw(polyline);
	}

	public void draw(double x, double y, double w, double h) {
		drawTriangle(x, y, w, h);
		
		if (w < .2 || h < .2)
			return;

		double halfW = w / 2;
		double halfH = h / 2;

		/*
			drawing order will be changed based on this
		*/
		draw(x, y, halfW, halfH); // left
		draw(x + halfW/2, y - halfH, halfW, halfH); // top
		draw(x + halfW, y, halfW, halfH); // right
	}

	public static void main(String[] args) {
		try {
			DrawFractal df = new DrawFractal();

			while (df.panel == null)
				TimeUnit.SECONDS.sleep(1);

			df.draw(0, 400, 400, 400);
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}
}

File: c++/blog/recursive/draw_fractal.cpp
#include <iostream>

using namespace std;

void drawTriangle(double x, double y, double w, double h)
{
	// {[x1, y1] - start point, [x2, y2] - right point, [x3, y3] - top point, [x1, y1] - end point, same as start point}
	int coordinates[] = {x, y, x+w, y, x+w/2, y+h, x, y};
}

void drawFractal(double x, double y, double w, double h)
{
	drawTriangle(x, y, w, h);

	if (w < .2 || h < .2)
		return;

	double halfH = h/2;
	double halfW = w/2;

	drawFractal(x, y, halfW, halfH); // left
	drawFractal(x + halfW/2, y + halfH, halfW, halfH); // top
	drawFractal(x + halfW, y, halfW, halfH); // right
}

void main()
{
	drawFractal(0, 0, 100, 100);
}
